_posts/2018-01-15-contributing-to-prettier-atom.md

```diff
@@ -0,0 +1,219 @@
+
+Atom is basically a web application, so it has all the same debug tooling as Chrome. The article also mentioned a `--profile-startup` flag that could be used to collect debug information for the initialization stage.
+
+I ran the following command in order to start atom and collect a profile of the startup phase.
+
+```bash
+atom --profile-startup .
+```
+
+## Analyzing profile information
+
+This is what the profile looked like in the Chrome javascript profiler.
+
+![Prettier javascript profile](/assets/posts/prettier-profile-full.png)
+
+The chart shows all the function calls that happen during initialization. The length of the bar indicates how long the function ran for. The layers represent functions called from within other functions.
+
+In other words, it's a [flame graph](http://www.brendangregg.com/FlameGraphs/cpuflamegraphs.html).
+
+By hovering over these bars, I was able to find the one corresponding to the `prettier-atom.activate` function and understand what was causing the slow activation time.
+
+Looking at the chart, a few things stood out.
+
+* Loading code adds up. Every `require` triggers a file to be read, parsed and compiled.
+* One file was requiring something called `toConsumableArray` which was taking 64ms to load.
+* Another file was requiring `lodash/fp` which was taking 111ms to run.
+
+![consumable array load time](/assets/posts/consumable-array-load-time.png)
+![lodash load time](/assets/posts/lodash-load-time.png)
+
+These were both occurring from a `require`. That means it wasn't time spent computing useful stuff, it was time spent simply loading the libraries. Yuck!
+
+If we could avoid loading these packages, we would save ~170ms from the activation time!
+
+## Optimization 1: Removing toConsumableArray
+
+I noticed that the `toConsumableArray` import was coming from `babel-runtime`. Babel runtime is a collection of polyfills to allow ES6 syntax in ES5 code. I looked at the file triggering the require, and sure enough there was some fancy syntax going on.
+
+```js
+const getAllScopes = () => [
+	...getJavascriptScopes(),
+	...getTypescriptScopes(),
+	...getCssScopes(),
+	...getJsonScopes(),
+	...getGraphQlScopes(),
+	...getMarkdownScopes()
+];
+```
+
+The `...` operator allows the arrays to be cleanly combined, but it's also not supported in ES5. So babel sees that syntax and adds `toConsumableArray` to polyfill the functionality.
+
+It's nice syntax, but not nice enough to warrant 64ms of activation time. Converting the expression to an ES5 compatible version stopped `babel-runtime` from being included.
+
+```js
+const getAllScopes = () =>
+	[
+		getJavascriptScopes(),
+		getTypescriptScopes(),
+		getCssScopes(),
+		getJsonScopes(),
+		getGraphQlScopes(),
+		getMarkdownScopes()
+	].reduce((allScopes, scopes) => allScopes.concat(scopes), []);
```

Comment:
Could this also just be `Array.prototype.concat(getJavascriptScopes(), ..., getMarkdownScopes())`? Maybe not super relevant to the point of the post.

Reply:
Oh wicked! I didn't think to look at whether `concat` took more than one parameter. 

Oh well... too late 😅